In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt

In [2]:
df1=pd.read_csv(r"D:\Big Data Analytics- Sem 3\Capstone Project\Toronto\Milestone1&2\Columns_to_attach_calendar.csv")
df2=pd.read_csv(r"D:\Big Data Analytics- Sem 3\Capstone Project\Toronto\Milestone1&2\clean_calendar.csv")

In [3]:
# List of columns to drop
columns_to_drop = ['available', 'adjusted_price', 'minimum_nights', 'maximum_nights']

# Drop the specified columns
df2 = df2.drop(columns=columns_to_drop)

In [4]:
# Concatenate 'accommodates', 'property', and 'neighbourhood' columns into 'identifier_listing'
df1['identifier_listing'] = df1.apply(lambda row: f"{row['accommodates']}_{row['neighbourhood']}", axis=1)

In [5]:
# Convert 'identifier_listing' column to 'category' dtype
df1['identifier_listing'] = df1['identifier_listing'].astype('category')

# Create a mapping dictionary
category_mapping = dict(enumerate(df1['identifier_listing'].cat.categories))

# Print the mapping dictionary
print(category_mapping)


{0: '1.0_0.0', 1: '1.0_1.0', 2: '1.0_10.0', 3: '1.0_100.0', 4: '1.0_101.0', 5: '1.0_102.0', 6: '1.0_103.0', 7: '1.0_104.0', 8: '1.0_105.0', 9: '1.0_106.0', 10: '1.0_107.0', 11: '1.0_108.0', 12: '1.0_109.0', 13: '1.0_11.0', 14: '1.0_110.0', 15: '1.0_111.0', 16: '1.0_112.0', 17: '1.0_113.0', 18: '1.0_114.0', 19: '1.0_115.0', 20: '1.0_116.0', 21: '1.0_117.0', 22: '1.0_118.0', 23: '1.0_119.0', 24: '1.0_12.0', 25: '1.0_120.0', 26: '1.0_121.0', 27: '1.0_122.0', 28: '1.0_123.0', 29: '1.0_124.0', 30: '1.0_125.0', 31: '1.0_126.0', 32: '1.0_127.0', 33: '1.0_128.0', 34: '1.0_129.0', 35: '1.0_13.0', 36: '1.0_130.0', 37: '1.0_131.0', 38: '1.0_132.0', 39: '1.0_133.0', 40: '1.0_134.0', 41: '1.0_135.0', 42: '1.0_136.0', 43: '1.0_137.0', 44: '1.0_138.0', 45: '1.0_139.0', 46: '1.0_14.0', 47: '1.0_15.0', 48: '1.0_16.0', 49: '1.0_17.0', 50: '1.0_18.0', 51: '1.0_19.0', 52: '1.0_2.0', 53: '1.0_20.0', 54: '1.0_21.0', 55: '1.0_22.0', 56: '1.0_23.0', 57: '1.0_24.0', 58: '1.0_25.0', 59: '1.0_26.0', 60: '1.0_27.

In [6]:
df1['identifier_listing'] = df1['identifier_listing'].astype('category').cat.codes

In [7]:
df1

accommodates  number_of_reviews  review_scores_rating  \
0                10                  6                  5.00   
1                 3                  8                  4.71   
2                 2                169                  4.84   
3                 4                  0                  0.00   
4                 4                 63                  4.93   
...             ...                ...                   ...   
21109             6                  0                  0.00   
21110             6                  0                  0.00   
21111             3                  0                  0.00   
21112             4                  0                  0.00   
21113             2                  0                  0.00   

       instant_bookable  calculated_host_listings_count  new_host  \
0                     0                               1         0   
1                     0                               1         0   
2                     0                               2         0   
3                     0                               1         0   
4                     0                               1         0   
...                 ...                             ...       ...   
21109                 0                               2         0   
21110                 0                               2         0   
21111                 0                               1         0   
21112                 1                               1         0   
21113                 0                               1         0   

       host_response_score  minimum_nights  maximum_nights  \
0                        0              28             730   
1                        8              28            1125   
2                        0             180             365   
3                        0              28              31   
4                       10             180            1125   
...                    ...             ...             ...   
21109                    0              28             365   
21110                    0              28             365   
21111                    0               2             365   
21112                    9               1             365   
21113                    0               1             365   

       minimum_nights_avg_ntm  maximum_nights_avg_ntm  availability_365  \
0                        28.0                   730.0                 0   
1                        28.0                  1125.0                19   
2                       180.0                   365.0                 0   
3                        28.0                    31.0                 0   
4                       180.0                  1125.0               345   
...                       ...                     ...               ...   
21109                    28.0                   365.0               269   
21110                    28.0                   365.0               269   
21111                     2.0                   365.0               351   
21112                     1.0                   365.0               204   
21113                     1.0                   365.0               250   

       host_score  property  neighbourhood  bathrooms  count_amenities  \
0               3        10             71         11              173   
1               2         6            122          2              145   
2               2        37            122          5              146   
3               2         6            132          2              141   
4               2        14             23          2              428   
...           ...       ...            ...        ...              ...   
21109           3        33              3          7              179   
21110           3        33              3          2              179   
21111           0        10             46          2              230   
21112           2      

In [8]:
# Merge df1 and df2 based on 'id' in df1 and 'listing_id' in df2
df = pd.merge(df1, df2, left_on='id', right_on='listing_id')



In [9]:
df

accommodates  number_of_reviews  review_scores_rating  \
0                  10                  6                   5.0   
1                  10                  6                   5.0   
2                  10                  6                   5.0   
3                  10                  6                   5.0   
4                  10                  6                   5.0   
...               ...                ...                   ...   
7706607             2                  0                   0.0   
7706608             2                  0                   0.0   
7706609             2                  0                   0.0   
7706610             2                  0                   0.0   
7706611             2                  0                   0.0   

         instant_bookable  calculated_host_listings_count  new_host  \
0                       0                               1         0   
1                       0                               1         0   
2                       0                               1         0   
3                       0                               1         0   
4                       0                               1         0   
...                   ...                             ...       ...   
7706607                 0                               1         0   
7706608                 0                               1         0   
7706609                 0                               1         0   
7706610                 0                               1         0   
7706611                 0                               1         0   

         host_response_score  minimum_nights  maximum_nights  \
0                          0              28             730   
1                          0              28             730   
2                          0              28             730   
3                          0              28             730   
4                          0              28             730   
...                      ...             ...             ...   
7706607                    0               1             365   
7706608                    0               1             365   
7706609                    0               1             365   
7706610                    0               1             365   
7706611                    0               1             365   

         minimum_nights_avg_ntm  ...  host_score  property  neighbourhood  \
0                          28.0  ...           3        10             71   
1                          28.0  ...           3        10             71   
2                          28.0  ...           3        10             71   
3                          28.0  ...           3        10             71   
4                          28.0  ...           3        10             71   
...                         ...  ...         ...       ...            ...   
7706607                     1.0  ...           3        33             33   
7706608                     1.0  ...           3        33             33   
7706609                     1.0  ...           3        33             33   
7706610                     1.0  ...           3        33             33   
7706611                     1.0  ...           3        33             33   

         bathrooms  count_amenities                   id  identifier_listing  \
0               11              173                 1419                 188   
1               11              173                 1419                 188   
2               11              173                 1419                 188   
3               11              173                 1419                 188   
4               11              173                 1419                 188   
...            ...              ...                  ...                 ...   
7706607          3              108  1175423299452956483                 377   
7706608          3           

In [10]:
df=df.drop(["id", "accommodates", "neighbourhood"],axis=1)

In [11]:
# Convert 'date' column to datetime dtype
df['date'] = pd.to_datetime(df['date'])

# Group by 'identifier_listing' and 'date' and calculate the mean of the other columns
df_1 = df.groupby(['identifier_listing', 'date']).mean().reset_index()



In [12]:
'''# Group by 'identifier_listing' and count the number of unique 'listing_id' values in each group
df3 = df2.groupby('identifier_listing')['listing_id'].nunique().reset_index()

# Filter for groups where the count of unique 'listing_id' values is greater than 1
filtered_df = df3[df3['listing_id'] > 1]

# Select the 'listing_id' and 'identifier_listing' columns from the filtered DataFrame
result_df = df2[df2['identifier_listing'].isin(filtered_df['identifier_listing'])]

# Print the 'listing_id' and 'identifier_listing' columns
print(result_df[['listing_id', 'identifier_listing']])'''


"# Group by 'identifier_listing' and count the number of unique 'listing_id' values in each group\ndf3 = df2.groupby('identifier_listing')['listing_id'].nunique().reset_index()\n\n# Filter for groups where the count of unique 'listing_id' values is greater than 1\nfiltered_df = df3[df3['listing_id'] > 1]\n\n# Select the 'listing_id' and 'identifier_listing' columns from the filtered DataFrame\nresult_df = df2[df2['identifier_listing'].isin(filtered_df['identifier_listing'])]\n\n# Print the 'listing_id' and 'identifier_listing' columns\nprint(result_df[['listing_id', 'identifier_listing']])"

In [13]:
df_1=df_1.drop("listing_id", axis=1)

In [14]:
df_1

identifier_listing       date  number_of_reviews  \
0                        0 2024-06-10          43.666667   
1                        0 2024-06-11          27.785714   
2                        0 2024-06-12          27.785714   
3                        0 2024-06-13          27.785714   
4                        0 2024-06-14          27.785714   
...                    ...        ...                ...   
445867                1220 2025-06-06           1.000000   
445868                1220 2025-06-07           1.000000   
445869                1220 2025-06-08           1.000000   
445870                1220 2025-06-09           1.000000   
445871                1220 2025-06-10           1.000000   

        review_scores_rating  instant_bookable  \
0                   4.780000          0.333333   
1                   3.007857          0.142857   
2                   3.007857          0.142857   
3                   3.007857          0.142857   
4                   3.007857          0.142857   
...                      ...               ...   
445867              5.000000          0.000000   
445868              5.000000          0.000000   
445869              5.000000          0.000000   
445870              5.000000          0.000000   
445871              5.000000          0.000000   

        calculated_host_listings_count  new_host  host_response_score  \
0                             2.000000  0.000000             9.833333   
1                             5.428571  0.142857             7.428571   
2                             5.428571  0.142857             7.428571   
3                             5.428571  0.142857             7.428571   
4                             5.428571  0.142857             7.428571   
...                                ...       ...                  ...   
445867                        1.000000  0.000000             0.000000   
445868                        1.000000  0.000000             0.000000   
445869                        1.000000  0.000000             0.000000   
445870                        1.000000  0.000000             0.000000   
445871                        1.000000  0.000000             0.000000   

        minimum_nights  maximum_nights  minimum_nights_avg_ntm  \
0            14.833333      322.500000               14.833333   
1            19.571429      322.857143               19.571429   
2            19.571429      322.857143               19.571429   
3            19.571429      322.857143               19.571429   
4            19.571429      322.857143               19.571429   
...                ...             ...                     ...   
445867       28.000000       36.000000               28.000000   
445868       28.000000       36.000000               28.000000   
445869       28.000000       36.000000               28.000000   
445870       28.000000       36.000000               28.000000   
445871       28.000000       36.000000               28.000000   

        maximum_nights_avg_ntm  availability_365  host_score   property  \
0                   871.666667             186.0    3.166667  33.000000   
1                   558.214286             143.0    2.714286  32.714286   
2                   558.214286             143.0    2.714286  32.714286   
3                   558.214286             143.0    2.714286  32.714286   
4                   558.214286             143.0    2.714286  32.714286   
...                        ...               ...         ...        ...   
445867               36.000000              90.0    1.000000  33.000000   
445868               36.000000              90.0    1.000000  33.000000   
445869               36.000000              90.0    1.000000  33.000000   
445870               36.000000              90.0    1.000000  33.000000   
445871               36.000000              90.0    1.000000  33.000000   

        bathrooms  count_amenities      price  
0        3.333333       457.500000  41.833333  
1        3.357143  

In [15]:
df_1.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 445872 entries, 0 to 445871
Data columns (total 18 columns):
 #   Column                          Non-Null Count   Dtype         
---  ------                          --------------   -----         
 0   identifier_listing              445872 non-null  int16         
 1   date                            445872 non-null  datetime64[ns]
 2   number_of_reviews               445872 non-null  float64       
 3   review_scores_rating            445872 non-null  float64       
 4   instant_bookable                445872 non-null  float64       
 5   calculated_host_listings_count  445872 non-null  float64       
 6   new_host                        445872 non-null  float64       
 7   host_response_score             445872 non-null  float64       
 8   minimum_nights                  445872 non-null  float64       
 9   maximum_nights                  445872 non-null  float64       
 10  minimum_nights_avg_ntm          445872 non-null  float64

In [16]:
from pmdarima import auto_arima

# Assuming df is your DataFrame and 'price' is the column with the prices
model = auto_arima(df_1['price'], trace=True, error_action='ignore', suppress_warnings=True)
model.fit(df_1['price'])


Performing stepwise search to minimize aic
 ARIMA(2,1,2)(0,0,0)[0] intercept   : AIC=4348135.525, Time=9.50 sec
 ARIMA(0,1,0)(0,0,0)[0] intercept   : AIC=4348250.556, Time=4.78 sec
 ARIMA(1,1,0)(0,0,0)[0] intercept   : AIC=4348129.560, Time=6.18 sec
 ARIMA(0,1,1)(0,0,0)[0] intercept   : AIC=4348129.525, Time=6.93 sec
 ARIMA(0,1,0)(0,0,0)[0]             : AIC=4348248.556, Time=3.21 sec
 ARIMA(1,1,1)(0,0,0)[0] intercept   : AIC=4348131.525, Time=8.47 sec
 ARIMA(0,1,2)(0,0,0)[0] intercept   : AIC=4348131.525, Time=10.35 sec
 ARIMA(1,1,2)(0,0,0)[0] intercept   : AIC=4348133.525, Time=11.88 sec
 ARIMA(0,1,1)(0,0,0)[0]             : AIC=4348127.525, Time=4.46 sec
 ARIMA(1,1,1)(0,0,0)[0]             : AIC=4348129.525, Time=5.22 sec
 ARIMA(0,1,2)(0,0,0)[0]             : AIC=4348129.525, Time=8.61 sec
 ARIMA(1,1,0)(0,0,0)[0]             : AIC=4348127.560, Time=4.29 sec
 ARIMA(1,1,2)(0,0,0)[0]             : AIC=4348131.525, Time=6.86 sec

Best model:  ARIMA(0,1,1)(0,0,0)[0]          
Total fit t

ARIMA(order=(0, 1, 1), scoring_args={}, suppress_warnings=True,
      with_intercept=False)

Output:     
Performing stepwise search to minimize aic
 ARIMA(2,1,2)(0,0,0)[0] intercept   : AIC=4348135.525, Time=9.50 sec
 ARIMA(0,1,0)(0,0,0)[0] intercept   : AIC=4348250.556, Time=4.78 sec
 ARIMA(1,1,0)(0,0,0)[0] intercept   : AIC=4348129.560, Time=6.18 sec
 ARIMA(0,1,1)(0,0,0)[0] intercept   : AIC=4348129.525, Time=6.93 sec
 ARIMA(0,1,0)(0,0,0)[0]             : AIC=4348248.556, Time=3.21 sec
 ARIMA(1,1,1)(0,0,0)[0] intercept   : AIC=4348131.525, Time=8.47 sec
 ARIMA(0,1,2)(0,0,0)[0] intercept   : AIC=4348131.525, Time=10.35 sec
 ARIMA(1,1,2)(0,0,0)[0] intercept   : AIC=4348133.525, Time=11.88 sec
 ARIMA(0,1,1)(0,0,0)[0]             : AIC=4348127.525, Time=4.46 sec
 ARIMA(1,1,1)(0,0,0)[0]             : AIC=4348129.525, Time=5.22 sec
 ARIMA(0,1,2)(0,0,0)[0]             : AIC=4348129.525, Time=8.61 sec
 ARIMA(1,1,0)(0,0,0)[0]             : AIC=4348127.560, Time=4.29 sec
 ARIMA(1,1,2)(0,0,0)[0]             : AIC=4348131.525, Time=6.86 sec

Best model:  ARIMA(0,1,1)(0,0,0)[0]          
Total fit time: 90.865 seconds


In [17]:
forecast = model.predict(n_periods=30)


In [18]:
forecast


445872    39.0
445873    39.0
445874    39.0
445875    39.0
445876    39.0
445877    39.0
445878    39.0
445879    39.0
445880    39.0
445881    39.0
445882    39.0
445883    39.0
445884    39.0
445885    39.0
445886    39.0
445887    39.0
445888    39.0
445889    39.0
445890    39.0
445891    39.0
445892    39.0
445893    39.0
445894    39.0
445895    39.0
445896    39.0
445897    39.0
445898    39.0
445899    39.0
445900    39.0
445901    39.0
dtype: float64

Output:     
445872    39.0
445873    39.0
445874    39.0
445875    39.0
445876    39.0
445877    39.0
445878    39.0
445879    39.0
445880    39.0
445881    39.0
445882    39.0
445883    39.0
445884    39.0
445885    39.0
445886    39.0
445887    39.0
445888    39.0
445889    39.0
445890    39.0
445891    39.0
445892    39.0
445893    39.0
445894    39.0
445895    39.0
445896    39.0
445897    39.0
445898    39.0
445899    39.0
445900    39.0
445901    39.0
dtype: float64

In [19]:
from statsmodels.tsa.arima.model import ARIMA
from sklearn.metrics import mean_squared_error, mean_absolute_error, mean_absolute_percentage_error
import numpy as np

# Assuming df is your DataFrame and 'price' is the column with the prices
train_size = int(len(df_1) * 0.8)
train, test = df_1[0:train_size], df_1[train_size:]

model = ARIMA(train['price'], order=(0,1,1))
model_fit = model.fit()

forecast = model_fit.predict(start=len(train), end=len(train)+len(test)-1)

# Evaluate the performance of the model on the test set
mse = mean_squared_error(test['price'], forecast)
rmse = np.sqrt(mse)
mae = mean_absolute_error(test['price'], forecast)
mape = mean_absolute_percentage_error(test['price'], forecast)

print('MSE:', mse)
print('RMSE:', rmse)
print('MAE:', mae)
print('MAPE:', mape)


MSE: 243600.49843854617
RMSE: 493.5590121135933
MAE: 264.7893994078712
MAPE: 0.4853656763795618


Output:     
MSE: 243600.49843854617
RMSE: 493.5590121135933
MAE: 264.7893994078712
MAPE: 0.4853656763795618


In [46]:
'''from keras.models import Sequential
from keras.layers import LSTM, Dense, Dropout
from sklearn.preprocessing import MinMaxScaler
from sklearn.metrics import mean_squared_error, mean_absolute_error, mean_absolute_percentage_error
import numpy as np

# Assuming df is your DataFrame and 'price' is the column with the prices
# Prepare the data for LSTM
scaler = MinMaxScaler(feature_range=(0, 1))
scaled_price = scaler.fit_transform(df_1['price'].values.reshape(-1,1))
scaled_amenities = scaler.fit_transform(df_1['count_amenities'].values.reshape(-1,1))
scaled_response_score = scaler.fit_transform(df_1['host_response_score'].values.reshape(-1,1))
scaled_host_score = scaler.fit_transform(df_1['host_score'].values.reshape(-1,1))
scaled_identifier = scaler.fit_transform(df_1['identifier_listing'].values.reshape(-1,1))

# Concatenate the scaled data along the column axis
scaled_data = np.concatenate((scaled_price, scaled_amenities, scaled_response_score, scaled_host_score, scaled_identifier), axis=1)

# Create sequences of data points with a fixed length
sequence_length = 30
X = []
y = []
for i in range(len(scaled_data)-sequence_length):
    X.append(scaled_data[i:i+sequence_length])
    y.append(scaled_data[i+sequence_length, 0])
X = np.array(X)
y = np.array(y)

# Split the data into training and test sets
train_size = int(len(X) * 0.8)
X_train, X_test = X[0:train_size], X[train_size:]
y_train, y_test = y[0:train_size], y[train_size:]

# Build the LSTM model
model = Sequential()
model.add(LSTM(units=50, return_sequences=True, input_shape=(X_train.shape[1], X_train.shape[2])))
model.add(Dropout(0.2))
model.add(LSTM(units=50, return_sequences=False))
model.add(Dropout(0.2))
model.add(Dense(units=1))

# Compile the model
model.compile(optimizer='adam', loss='mean_squared_error')

# Train the model
model.fit(X_train, y_train, epochs=5, batch_size=32)

# Make predictions on the test set
predictions = model.predict(X_test)
predictions = scaler.inverse_transform(predictions)

# Evaluate the performance of the model on the test set
mse = mean_squared_error(y_test, predictions)
rmse = np.sqrt(mse)
mae = mean_absolute_error(y_test, predictions)
mape = mean_absolute_percentage_error(y_test, predictions)

print('MSE:', mse)
print('RMSE:', rmse)
print('MAE:', mae)
print('MAPE:', mape)'''


c:\Users\vidha\AppData\Local\Programs\Python\Python311\Lib\site-packages\keras\src\layers\rnn\rnn.py:204: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(**kwargs)


Epoch 1/5
11147/11147 ━━━━━━━━━━━━━━━━━━━━ 147s 13ms/step - loss: 2.9510e-04
Epoch 2/5
11147/11147 ━━━━━━━━━━━━━━━━━━━━ 164s 15ms/step - loss: 6.9556e-05
Epoch 3/5
11147/11147 ━━━━━━━━━━━━━━━━━━━━ 161s 14ms/step - loss: 5.3570e-05
Epoch 4/5
11147/11147 ━━━━━━━━━━━━━━━━━━━━ 166s 15ms/step - loss: 4.6232e-05
Epoch 5/5
11147/11147 ━━━━━━━━━━━━━━━━━━━━ 161s 14ms/step - loss: 4.0698e-05
2787/2787 ━━━━━━━━━━━━━━━━━━━━ 12s 4ms/step
MSE: 6780.422767326051
RMSE: 82.3433225424263
MAE: 64.62290447319616
MAPE: 1656.8346772500504
